Mathlib/Algebra/Group/Equiv/Basic.lean

```diff
@@ -80,19 +81,91 @@ def arrowCongr {M N P Q : Type*} [Mul P] [Mul Q] (f : M ≃ N) (g : P ≃* Q) :
   right_inv k := by ext; simp
   map_mul' h k := by ext; simp
 
+section monoidHomCongrEquiv
+variable [MulOneClass M] [MulOneClass M₁] [MulOneClass M₂] [MulOneClass M₃]
+  [Monoid N] [Monoid N₁] [Monoid N₂] [Monoid N₃]
+
+/-- The equivalence `(M₁ →* N) ≃ (M₂ →* N)` obtained by postcomposition with
+a multiplicative equivalence `e : M₁ ≃* M₂`. -/
+@[to_additive (attr := simps)
+"The equivalence `(M₁ →+ N) ≃ (M₂ →+ N)` obtained by postcomposition with
+an additive equivalence `e : M₁ ≃+ M₂`."]
+def monoidHomCongrLeftEquiv (e : M₁ ≃* M₂) : (M₁ →* N) ≃ (M₂ →* N) where
+  toFun f := f.comp e.symm.toMonoidHom
+  invFun f := f.comp e.toMonoidHom
+  left_inv f := by ext; simp
+  right_inv f := by ext; simp
+
+/-- The equivalence `(M →* N₁) ≃ (M →* N₂)` obtained by postcomposition with
+a multiplicative equivalence `e : N₁ ≃* N₂`. -/
+@[to_additive (attr := simps)
+"The equivalence `(M →+ N₁) ≃ (M →+ N₂)` obtained by postcomposition with
+an additive equivalence `e : N₁ ≃+ N₂`."]
+def monoidHomCongrRightEquiv (e : N₁ ≃* N₂) : (M →* N₁) ≃ (M →* N₂) where
+  toFun := e.toMonoidHom.comp
+  invFun := e.symm.toMonoidHom.comp
+  left_inv f := by ext; simp
+  right_inv f := by ext; simp
+
+@[to_additive (attr := simp)]
+lemma monoidHomCongrLeftEquiv_refl : monoidHomCongrLeftEquiv (.refl M) = .refl (M →* N) := rfl
+
+@[to_additive (attr := simp)]
+lemma monoidHomCongrRightEquiv_refl : monoidHomCongrRightEquiv (.refl N) = .refl (M →* N) := rfl
+
+@[to_additive (attr := simp)]
+lemma monoidHomCongrLeftEquiv_symm (e : M₁ ≃* M₂) :
+    monoidHomCongrLeftEquiv (N := N) e.symm = (monoidHomCongrLeftEquiv e).symm := rfl
+
+@[to_additive (attr := simp)]
+lemma monoidHomCongrRightEquiv_symm (e : N₁ ≃* N₂) :
+    monoidHomCongrRightEquiv (M := M) e.symm = (monoidHomCongrRightEquiv e).symm := rfl
```

Comment:
I'd normally argue this should go the other way, to push the `symm` in through the functorial equivs until it hits the interesting one (like `toAdd.symm = ofAdd`)

Reply:
Yes, you're right. The other way, it's a projection lemma